ducktape/services/service.py

```diff
@@ -173,12 +176,33 @@ def start_node(self, node):
         """Start service process(es) on the given node."""
         raise NotImplementedError("%s: subclasses must implement start_node." % self.who_am_i())
 
-    def wait(self):
+    def wait(self, timeout_sec=600):
         """Wait for the service to finish.
         This only makes sense for tasks with a fixed amount of work to do. For services that generate
         output, it is only guaranteed to be available after this call returns.
         """
-        pass
+        alive_nodes = [] # nodes still alive following completion of wait
+        start = time.time()
+        end = start + timeout_sec
+        for node in self.nodes:
+            now = time.time()
+            if end > now:
+                self.logger.debug("%s: waiting for node", self.who_am_i(node))
+                if not self.wait_node(node, end - now):
+                    alive_nodes.append(node)
+            else:
+                alive_nodes.append(node)
+
+        if alive_nodes:
+            raise TimeoutError("Timed out waiting %s seconds for background threads to finish. " % str(timeout_sec) +
```

Comment:
Make this message slightly more generic now that we've pulled it up to `Service`?